client/core/core.go

```diff
@@ -113,10 +151,308 @@ func (c *Core) ListMarkets() []*MarketInfo {
 	return infos
 }
 
-func (c *Core) Register(*Registration) error {
+// wallet gets the wallet for the specified asset ID in a thread-safe way.
+func (c *Core) wallet(assetID uint32) (*xcWallet, bool) {
+	c.walletMtx.RLock()
+	defer c.walletMtx.RUnlock()
+	w, found := c.wallets[assetID]
+	return w, found
+}
+
+func (c *Core) Wallets() []*WalletStatus {
+	c.walletMtx.RLock()
+	defer c.walletMtx.RUnlock()
+	stats := make([]*WalletStatus, 0, len(c.wallets))
+	for assetID, wallet := range c.wallets {
+		on, open := wallet.status()
+		stats = append(stats, &WalletStatus{
+			AssetID: assetID,
+			Symbol:  unbip(assetID),
+			Open:    open,
+			Running: on,
+		})
+	}
+	return stats
+}
+
+// CreateWallet creates a new exchange wallet.
+func (c *Core) CreateWallet(form *WalletForm) error {
+	dbWallet := &db.Wallet{
+		AssetID: form.AssetID,
+		Account: form.Account,
+		INIPath: form.INIPath,
+	}
+	_, exists := c.wallet(form.AssetID)
+	if exists {
+		return fmt.Errorf("%s wallet", unbip(dbWallet.AssetID))
+	}
+	wallet := &xcWallet{AssetID: form.AssetID}
+
+	walletCfg := &asset.WalletConfig{
+		Account: dbWallet.Account,
+		INIPath: dbWallet.INIPath,
+		TipChange: func(err error) {
+			c.tipChange(form.AssetID, err)
+		},
+	}
+
+	logger := c.loggerMaker.SubLogger("CORE", unbip(dbWallet.AssetID))
+	w, err := asset.Setup(dbWallet.AssetID, walletCfg, logger, c.net)
+	if err != nil {
+		return fmt.Errorf("error creating wallet: %v", err)
+	}
+	wallet.Wallet = w
+	wallet.waiter = dex.NewStartStopWaiter(w)
+	wallet.waiter.Start(c.ctx)
+	err = wallet.Connect()
+	if err != nil {
+		return fmt.Errorf("Error connecting wallet: %v", err)
+	}
+
+	// Store the wallet in the database.
+	err = c.db.UpdateWallet(dbWallet)
+	if err != nil {
+		wallet.waiter.Stop()
+		return fmt.Errorf("error storing wallet credentials: %v", err)
+	}
+
+	c.walletMtx.Lock()
+	c.wallets[form.AssetID] = wallet
+	c.walletMtx.Unlock()
+
 	return nil
 }
 
+// WalletStatus returns 1) whether the wallet exists and 2) if it's currently
+// running.
+func (c *Core) WalletStatus(assetID uint32) (has, running, open bool) {
+	c.walletMtx.Lock()
+	defer c.walletMtx.Unlock()
+	wallet, has := c.wallets[assetID]
+	if !has {
+		return
+	}
+	running, open = wallet.status()
+	return
+}
+
+// OpenWallet opens the wallet for use.
+func (c *Core) OpenWallet(assetID uint32, pw string) error {
+	wallet, found := c.wallet(assetID)
+	if !found {
+		return fmt.Errorf("no wallet for %d -> %s", assetID, unbip(assetID))
+	}
+	return wallet.Unlock(pw, aYear)
+}
+
+// Register registers an account with a new DEX. Register will block for the
+// entire process, including waiting for confirmations.
+func (c *Core) Register(form *Registration) (error, <-chan error) {
```

Comment:
Where is `form.Password` going to be sourced from? From the looks of things I'd say a password gets generated randomly and stored in the db and field in core. This way every dex account generated can use the same core password. Let me know if you're thinking of something different.

Reply:
Password is user-supplied and unique to the DEX right now,

https://github.com/buck54321/dcrdex/blob/e73d16f774f9d8fe690e0dc5e9655b68ca052a41/client/webserver/api.go#L33

but I agree that a single common password would be better. Follow-up, though.